### ISSUE TITLE
App language can't be changed

### ISSUE THREAD

**Issue by cremor** [webosbrew/youtube-webos#117]
The app was automatically set to my home country and native language after installing it. I want to change the app's language to English. But I can't change the language in the app's settings. After I choose a new language a dialog screen is shown where I have to confirm an app restart to change the language. But that confirm button doesn't do anything.

App version 0.3.1 (installed via Homebrew Channel)
OLED C9
Firmware 05.30.40
webOS 4.9.7
TV is **not** rooted.

**Comment by throwaway96** [comment]
Seems to work for me on webOS 6.3.2.

**Comment by cremor** [comment]
That's weird. I've also just tried it with your test version v0.3.3-dev3 and it still doesn't work for me.

I don't know if it helps, but the app shows the follwing information:

> web_20240325_10_RC00
> TV (PLATFORM_DETAIL_TV), LG/OLED65C9PLA/O18, LG Browser (8.00.00), webOS (2019)
> TVHTML5, 7.20240325.10.00

**Comment by puppinoo** [comment]
This happens for me too.
Installed lates beta on a freshly rooted LGC9.

Thanks.

**Comment by cremor** [comment]
> Seems to work for me on webOS 6.3.2.

@throwaway96 Could this be related to the user agent that the app is using? I noticed that the user agent is build dynamically (see https://github.com/webosbrew/youtube-webos/blob/main/assets/appinfo.json#L17), so you will have a newer browser version in your user agent than me. And according to https://github.com/reisxd/TizenTube/issues/74 YouTube uses the user agent to change its behavior. (The CSP problems mentioned in that issue shouldn't apply to WebOS because of the `allowCrossDomain` setting which is available in WebOS.)

**Comment by cremor** [comment]
Thanks to @fire332 I now know how to debug the app. I can see that an HTTP 400 error is happening when I try to confirm the new language in the dialog screen.

Below is the info from the network tab.
Does anyone know how a correct request would have to look like?

```
General
Request URL:https://www.youtube.com/tv_config?action_update_language=1
Request Method:POST
Status Code:400 
Remote Address:[2a00:1450:4014:80b::200e]:443

Reponse Headers
alt-svc:h3=":443"; ma=2592000,h3-29=":443"; ma=2592000
cache-control:no-cache, no-store, max-age=0, must-revalidate
content-encoding:gzip
content-type:text/html; charset=utf-8
date:Fri, 03 Jan 2025 12:32:26 GMT
expires:Mon, 01 Jan 1990 00:00:00 GMT
pragma:no-cache
server:ESF
status:400
strict-transport-security:max-age=31536000
x-content-type-options:nosniff
x-frame-options:SAMEORIGIN
x-xss-protection:0

Request Headers
Provisional headers are shown
Accept-Language:de-AT, en;q=0.8, *;q=0.6
Content-Type:application/x-www-form-urlencoded;charset=UTF-8
Origin:https://www.youtube.com
Referer:https://www.youtube.com/tv?env_forceFullAnimation=1&env_enableWebSpeech=1&env_enableVoice=1
User-Agent:LG Browser/8.00.00 (webOS.TV-2019), _TV_O18/05.40.61 (LG, OLED65C9PLA, wired)

Query String Parameters
action_update_language:1

Form Data
session_token:<removed>
hl:en
```

**Comment by fire332** [comment]
> Provisional headers are shown
This hints at a CORS violation. Is there a network request right before it labeled "preflight"?

**Comment by cremor** [comment]
No. That "Provisional headers are shown" text is also shown for all successful requests.

**Comment by cremor** [comment]
@fire332 Does changing the language work for you? If so, could you please compare your request to mine?

**Comment by fire332** [comment]
I actually still have the email notification for this issue saved in my inbox 😅. I've just been much too busy with life this past 3 months but I'm not ignoring this issue. 

@Crazor you can try checking the network tab on the regular website version of YouTube for the request that occurs when changing the language and comparing the successful request/response to the one causing a 400 client error.

**Comment by cremor** [comment]
By overriding the user agent you can get https://www.youtube.com/tv running on a desktop browser. There the language change works. It logged this request:

```
POST /tv_config?action_update_language=1 HTTP/2
Host: www.youtube.com
User-Agent: Microsoft Edge (EdgeHTML) - Xbox // NOTE: USED JUST FOR TESTING ON DESKTOP
Accept: */*
Accept-Language: de,en-US;q=0.7,en;q=0.3
Accept-Encoding: gzip, deflate, br, zstd
Content-Type: application/x-www-form-urlencoded;charset=utf-8
Content-Length: 232
Referer: https://www.youtube.com/tv?hrld=2&fltor=1
Origin: https://www.youtube.com
DNT: 1
Sec-GPC: 1
Sec-Fetch-Dest: empty
Sec-Fetch-Mode: cors
Sec-Fetch-Site: same-origin
Connection: keep-alive
Alt-Used: www.youtube.com
Cookie: REPLACED
Priority: u=0
Pragma: no-cache
Cache-Control: no-cache
TE: trailers

POST Content:
session_token=REPLACED
hl=en

```

At a first glance this looks similar to the failed request I posted above. But the `Sec` headers are new. So maybe indeed a CORS issue since the TV app doesn't set those headers?

**Comment by cremor** [comment]
I've now tested it with Chrome 79 on the desktop, because this is the version that the TV uses. With that browser it also does **not** work on the desktop. So it looks like this problem is caused by the old Chrome version?

That browser sets the `sec-fetch-mode` and `sec-fetch-site` headers, but not `Sec-GPC` or `Sec-Fetch-Dest`. I wanted to try to add those headers but couldn't get custom fetch requests to work (not even the one which is successful in Firefox).

Here are the headers from the failing request with Chrome 79 on desktop:

```
POST https://www.youtube.com/tv_config?action_update_language=1
Request headers:
accept: */*
accept-encoding: gzip, deflate, br
accept-language: de-DE,de;q=0.9,en-US;q=0.8,en;q=0.7
content-length: 232
content-type: application/x-www-form-urlencoded;charset=UTF-8
cookie: REPLACED
origin: https://www.youtube.com
referer: https://www.youtube.com/tv?is_account_switch=1&hrld=1
sec-fetch-mode: cors
sec-fetch-site: same-origin
user-agent: LG Browser/8.00.00 (webOS.TV-2019), _TV_O18/05.40.61 (LG, OLED65C9PLA, wired)
```

**Comment by fire332** [comment]
[Here's my network tab info.](https://github.com/user-attachments/files/20742858/info.txt) I don't see anything that stands out. If you want to manually add the headers, you need to set a breakpoint right before it calls the `send` function and modify it there.

![Image](https://github.com/user-attachments/assets/c9b83078-ddd3-4905-b134-32bdad2d002a)

![Image](https://github.com/user-attachments/assets/c46aed4e-3908-4def-80bb-88c7dc6103cb)

![Image](https://github.com/user-attachments/assets/72df4561-1775-490a-8d2d-33a381875d4a)

**Comment by cremor** [comment]
Thank you so much for testing this and sharing the response!

This indeed looks very similar. I don't understand why YouTube accepts the request from your TV, but not mine.
Maybe there is something different with the old Chrome version that is not visible in headers?

Anyway, before playing around with the headers I noticed that your response contained a `set-cookie` header. So I simply tried to set this cookie on my TV via the developer console by executing this command:
`document.cookie = "PREF=f6=80&hl=en; Domain=.youtube.com; Expires=Fri, 13-Feb-2035 11:11:29 GMT; Path=/; Secure";`

And after a restart of the app it works!

Maybe a custom language selection option would be simpler than fixing the YouTube app? I assume most users would either want their native (TV) language, or english. So maybe a simple "override app language to english" option would be enough? That option could set the cookie if checked, and clear it if unchecked.

**Comment by cremor** [comment]
TizenTube - a similar custom YouTube app for Samsung TVs - fixed this by patching the `resolveCommand` function. Maybe the same fix could be applied here?
See reisxd/TizenTube#177 and https://github.com/reisxd/TizenTube/commit/522f83cc012d5b1c75181b651e084a3afa280323

**Comment by fire332** [comment]
Try out #316.

IPK Download: https://github.com/webosbrew/youtube-webos/actions/runs/16612628730/artifacts/3644949374
